### PR TITLE
[YUNIKORN-3329] Reject task-group placeholder resources that overflow or are negative

### PR DESCRIPTION
**Pull request by thc1006**
### Description

`GetTGResource` (`pkg/common/resource.go`) computed the gang placeholder ask as `members * minResource` with unchecked `int64` accessors, and `setTaskGroups` aggregates the per-group results with the unchecked `common.Add`. Because `minMember` and `minResource` come unbounded from the `yunikorn.apache.org/task-groups` annotation, the ask could wrap to a negative value in three places: the `Quantity.MilliValue()`/`Value()` accessor, the `members * value` product, and the cross-task-group aggregate. That negative value is then sent to the core as `AddApplicationRequest.PlaceholderAsk`.

Rather than saturate (which would hide an invalid annotation as a near-infinite gang reservation), this rejects the annotation at the parse boundary. `GetTaskGroupsFromAnnotation` now returns an error when a `minResource` is negative, when a cpu/other quantity's int64 accessor would overflow, when `minMember * minResource` would overflow, or when the aggregate across task groups would overflow.

The aggregate is checked in the scheduler-interface canonical key-space that `GetTGResource` actually emits (an implicit `pods` count per task group, and `cpu` mapped to `siCommon.CPU`/`vcore`), with same-canonical-key collisions rejected, so the check matches the real placeholder ask rather than the raw annotation keys.

Behaviorally this is the existing invalid-annotation path (YUNIKORN-1089): the task-group definition is discarded and the application is scheduled without gang reservation rather than being rejected outright. The change only ensures a wrapped/negative placeholder ask is not sent to the core.

### Type of change

- [x] Bug Fix

### Jira issue

Jira ID : https://issues.apache.org/jira/browse/YUNIKORN-3329

- [x] I have created a Jira issue for this pull request.
- [x] The Jira ID is part of the title of this pull request.

### AI Tooling
If an AI tool was used:
- [ ] The PR includes the phrase "Generated by \<tool>", where \<tool> is the name of the AI tool used.
- [ ] My use of AI contributions follows the ASF legal policy.

Check https://www.apache.org/legal/generative-tooling.html for details.

### How has this been tested?

- [x] New unit tests were added to cover new or changed code paths.
- [x] `make test_all` was run, and no failures reported.
- [ ] `make e2e_test` was run, and no failures reported.
- [ ] new e2e tests have been added.

`TestGetTaskGroupsFromAnnotationValidatesResources` rejects a negative cpu, an int64-accessor overflow (cpu `MilliValue` and memory `Value`), a `minMember * minResource` product overflow, and an aggregate overflow, including the `cpu`/`vcore` canonical collision and the explicit/implicit `pods` collision, and accepts a normal task group. `TestValidatedTaskGroupsNeverProduceNegativePlaceholderAsk` is a caller-level property test: for every annotation the validator accepts, it rebuilds the placeholder ask the way `setTaskGroups` does (`common.Add` over `GetTGResource`) and asserts every resource is non-negative. Both fail before the change and pass after. `make test_all` passes (lint reports 0 issues, license checks OK, all unit tests pass); e2e was not run.

### Questions:

- [ ] The change needs documentation, a pull request for apache/yunikorn-site repository will be created.
- [ ] There is breaking changes for older versions: jira is tagged with `release-notes` label.
- [ ] The licenses files needs to be updated.

### CODE REVIEW

**Comment by chia7712** [comment]
> Pending. ASF Jira account requested via https://selfserve.apache.org/jira-account.html; will link the YUNIKORN issue once approved.

I have approved your request. Please file a jira ticket for this PR. Thanks!

**Comment by thc1006** [comment]
> > Pending. ASF Jira account requested via https://selfserve.apache.org/jira-account.html; will link the YUNIKORN issue once approved.
> 
> I have approved your request. Please file a jira ticket for this PR. Thanks!

I just got my ID haha! ok I wil do it later !!! 
Get my breakfast first!!!

**Comment by chia7712** [comment]
> Get my breakfast first!!!

sure. Food comes first. No rush!

File: pkg/cache/utils.go [hunk 1]
```diff
@@ -58,5 +61,77 @@ func GetTaskGroupsFromAnnotation(pod *v1.Pod) ([]TaskGroup, error) {
 				taskGroupInfo)
 		}
 	}
+	if err := validateTaskGroupResources(taskGroups); err != nil {
+		return nil, err
+	}
```
Comment (wilfred-s) [inline pkg/cache/utils.go]: Instead of doing a new range can this not be included in the range above? Saves one loop
```
totals := make(map[string]int64)
for range taskGroups {
...
    err := validateTaskGroupResources(taskGroup, totals)
    if err != nil {
        return nil, err
    }
}
```
Check totals inside the validate call directly.

File: pkg/cache/utils.go [hunk 2]
```diff
@@ -58,5 +61,77 @@ func GetTaskGroupsFromAnnotation(pod *v1.Pod) ([]TaskGroup, error) {
...
+// actually emit: each task group contributes an implicit "pods" count plus one
+// entry per minResource under its scheduler-interface canonical name ("cpu" becomes
+// siCommon.CPU, i.e. "vcore"), and the groups are summed. minMember and minResource
+// come unbounded from the pod annotation, so an unchecked int64 accessor, a
+// minMember*minResource product, or the aggregate across task groups can wrap to a
+// negative value that is then sent to the core as PlaceholderAsk. Because "cpu"
+// canonicalizes to the same key as "vcore" and an explicit "pods" collides with the
+// implicit count, the aggregate is computed in the canonical key-space with
+// same-key collisions rejected, so the check matches the real placeholder ask.
+func validateTaskGroupResources(taskGroups []TaskGroup) error {
```
Comment (wilfred-s) [inline pkg/cache/utils.go]: Change signature, see above

File: pkg/cache/utils.go [hunk 3]
```diff
@@ -58,5 +61,77 @@ func GetTaskGroupsFromAnnotation(pod *v1.Pod) ([]TaskGroup, error) {
...
+	totals := make(map[string]int64)
+	for _, taskGroup := range taskGroups {
+		members := int64(taskGroup.MinMember)
+		// group mirrors the si.Resource GetTGResource emits for this task group.
+		group := map[string]int64{"pods": members}
```
Comment (wilfred-s) [inline pkg/cache/utils.go]: update totals directly with a check for overflow and return error on fail.

File: pkg/cache/utils.go [hunk 4]
```diff
@@ -58,5 +61,77 @@ func GetTaskGroupsFromAnnotation(pod *v1.Pod) ([]TaskGroup, error) {
...
+		for resName, quantity := range taskGroup.MinResource {
+			value, err := checkedResourceValue(resName, quantity, taskGroup.Name)
```
Comment (wilfred-s) [inline pkg/cache/utils.go]: * check sign here first, error on fail
* make canonical for CPU 
* calculate max value based on type (cpu/vcore or not) and members:
```
milliConvert := 1
if cpu {
	milliConvert = 1000
}
maxVal := math.MaxInt64/(members*milliConvert)
if quantity.CmpInt64(maxVal) > 0 {
 return err...
}
```
* update totals directly with a check for overflow and return error on fail

File: pkg/cache/utils.go [hunk 5]
```diff
@@ -58,5 +61,77 @@ func GetTaskGroupsFromAnnotation(pod *v1.Pod) ([]TaskGroup, error) {
...
+				return fmt.Errorf("minResource %q in taskGroup %q collides with canonical resource %q", resName, taskGroup.Name, canonical)
+			}
+			group[canonical] = members * value
+		}
+		for name, value := range group {
+			if totals[name] > math.MaxInt64-value {
+				return fmt.Errorf("aggregate placeholder request for %q overflows int64 across taskGroups", name)
+			}
+			totals[name] += value
+		}
```
Comment (wilfred-s) [inline pkg/cache/utils.go]: directly in the above loop

File: pkg/cache/utils.go [hunk 6]
```diff
@@ -58,5 +61,77 @@ func GetTaskGroupsFromAnnotation(pod *v1.Pod) ([]TaskGroup, error) {
...
+	}
+	return nil
+}
+
+// checkedResourceValue returns the int64 value GetTGResource uses for a
+// (resourceName, quantity) pair, rejecting a negative or non-int64-representable
+// quantity. cpu is measured in milli-units (MilliValue), everything else in whole
+// units (Value), matching GetTGResource.
+func checkedResourceValue(resName string, quantity resource.Quantity, groupName string) (int64, error) {
```
Comment (wilfred-s) [inline pkg/cache/utils.go]: This can be done directly in the loop above

**Review by wilfred-s — CHANGES_REQUESTED** [review]

**Comment by thc1006** [comment]
Thanks, this reads a lot cleaner. I've pushed the rework:
- Validation now runs inside the existing task-group loop, no second pass.
- `validateTaskGroupResources(taskGroup, totals)` takes the shared `totals` and updates it directly.
- Per resource: sign check first, then the cpu->vcore canonical, then your `maxVal := math.MaxInt64/(members*milliConvert)` with `CmpInt64`. Dropped `checkedResourceValue` and `maxMilliCPU`.

One place I didn't fold in: I kept a small per-group `seen` set that rejects a same-canonical-key collision (cpu together with vcore, or an explicit `pods`). GetTGResource adds the implicit `pods` first and then iterates the map, so an explicit `pods` deterministically overwrites that count, and cpu vs vcore overwrite each other by map order. Rejecting felt safer than passing a silently-overwritten ask to the core, but I'm happy to drop it and let them sum if you'd rather.

Smaller note on the cpu bound: `maxVal` floor-divides by 1000, so it conservatively rejects cpu in the sub-milli remainder above roughly 9e15 cores (`MilliValue` rounds up). That's fail-safe and well past anything real, so I left your simpler bound, but I can make it an exact Quantity limit if you want it precise.

I also strengthened the property test to check each accepted ask against an exact big.Int recomputation instead of just the sign, which pins the validator's canonicalization to GetTGResource's.

**Comment by thc1006** [comment]
@wilfred-s when you have a moment, could you take another look? The validation now runs inside the existing task-group loop, so the separate pass is gone. I also added a unit test for the pods aggregate-overflow guard. Thanks for the earlier review.

**Comment by thc1006** [comment]
Hi @wilfred-s, I've reworked this along the lines you suggested. The aggregation now happens inside the single pass over the task groups: `GetTaskGroupsFromAnnotation` calls `validateTaskGroupResources` once per group, and that checks the sign and the overflow inline, updates the running totals directly, and returns an error on failure. The separate second range is gone. I also rebased onto current master. Could you take another look when you have a chance?

**Review by wilfred-s — APPROVED** [review]
LGTM, waiting on the CI to finish before committing.
